### PR TITLE
[receiver/hostmetrics] Set process metrics start time to process create time

### PR DESCRIPTION
**Pull request by hsmatulis**
**Description:** The start time of process metrics is currently set to the boot time of the system. This means that if a user repeatedly runs the same command, eventually the user will get two processes with the same PID and command, with the same start time (the system boot time). This makes the two data points indistinguishable and creates an association where there is none.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/11447

**Testing:**  Ran related unit tests through `make test`

### CODE REVIEW

File: cmd/mdatagen/metrics_v2.tmpl [hunk 1]
```diff
@@ -302,3 +302,8 @@ func (mb *MetricsBuilder) Reset(options ...metricBuilderOption) {
 		op(mb)
 	}
 }
+
+// SetStartTime sets the start time of metrics produced by the metrics builder to the given value.
+func (mb *MetricsBuilder) SetStartTime(startTime pcommon.Timestamp) {
+	mb.startTime = startTime
+}
```
Comment (dmitryax) [inline cmd/mdatagen/metrics_v2.tmpl:309]: Please supply `WithStartTimeOverride` option to `EmitForResource` call instead of adding this method

Comment (hsmatulisgoogle) [reply cmd/mdatagen/metrics_v2.tmpl]: Done! This makes a lot more sense, I mistakenly thought it could only be used by the Emit function

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 2]
```diff
@@ -112,13 +111,17 @@ func (s *scraper) scrape(_ context.Context) (pmetric.Metrics, error) {
 		if err = s.scrapeAndAppendDiskIOMetric(now, md.handle); err != nil {
 			errs.AddPartial(diskMetricsLen, fmt.Errorf("error reading disk usage for process %q (pid %v): %w", md.executable.name, md.pid, err))
 		}
-
-		s.mb.EmitForResource(md.resourceOptions()...)
+		options := append([]metadata.ResourceMetricsOption{metadata.WithStartTimeOverride(pcommon.Timestamp(start_time * 1e9))}, md.resourceOptions()...)
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:114]: Just for better readability :)

```suggestion
		options := append(md.resourceOptions(), metadata.WithStartTimeOverride(pcommon.Timestamp(start_time * 1e9)))
```

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 3]
```diff
@@ -99,6 +93,11 @@ func (s *scraper) scrape(_ context.Context) (pmetric.Metrics, error) {
 	}
 
 	for _, md := range data {
+		start_time, err := s.getProcessCreateTime(md.handle)
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:96]: Please use camelCase

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 4]
```diff
@@ -112,13 +111,17 @@ func (s *scraper) scrape(_ context.Context) (pmetric.Metrics, error) {
...
+		s.mb.EmitForResource(options...)
 	}
 
 	return s.mb.Emit(), errs.Combine()
 }
 
+func WithStartTimeOverride() {
+	panic("unimplemented")
+}
+
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:124]: Why is this needed?

Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go]: This was included by accident

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 3 above)
Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go]: Done

**Comment by dmitryax** [comment]
Please fix the errors and add a changelog entry

**Comment by hsmatulisgoogle** [comment]
Done fixing errors and adding the changelog entry

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go [hunk 5]
```diff
@@ -280,6 +280,11 @@ func (p *processHandleMock) Username() (string, error) {
 	return args.String(0), args.Error(1)
 }
 
+func (p *processHandleMock) CreateTime() (int64, error) {
+	args := p.MethodCalled("CreateTime")
+	return args.Get(0).(int64), args.Error(1)
+}
+
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go:287]: Why is this moved?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go [hunk 6]
```diff
@@ -422,10 +423,9 @@ func TestScrapeMetrics_Filtered(t *testing.T) {
 
 			handles := make([]*processHandleMock, 0, len(test.names))
 			for i, name := range test.names {
-				handleMock := newDefaultHandleMock()
+				handleMock := newDefaultHandleMock(time.Now().UnixMilli() - test.upTimeMs[i])
 				handleMock.On("Name").Return(name, nil)
 				handleMock.On("Exe").Return(name, nil)
-				handleMock.On("CreateTime").Return(time.Now().UnixMilli()-test.upTimeMs[i], nil)
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go:428]: Is this change really needed or we can keep? Does it change anything?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go [hunk 7]
```diff
@@ -114,3 +115,8 @@ func getProcessHandlesInternal() (processHandles, error) {
 
 	return &gopsProcessHandles{handles: processes}, nil
 }
+
+func getProcessCreateTimeInternal(p processHandle) (uint64, error) {
+	v, e := p.CreateTime()
+	return uint64(v), e
+}
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:122]: `uint64` is still a not the format that is used eventually, right? We will need to get `pcommon.Timestamp` out of it anyway. Do we really need this wrapper? Can we use `p.CreateTime` instead?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go [hunk 8]
```diff
@@ -82,12 +83,12 @@ type processHandle interface {
 	Name() (string, error)
 	Exe() (string, error)
 	Username() (string, error)
+	CreateTime() (int64, error)
 	Cmdline() (string, error)
 	CmdlineSlice() ([]string, error)
 	Times() (*cpu.TimesStat, error)
 	MemoryInfo() (*process.MemoryInfoStat, error)
 	IOCounters() (*process.IOCountersStat, error)
-	CreateTime() (int64, error)
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:90]: Why moved?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go (hunk 7 above)
Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:145]: An `uint64` output makes the `s.scrapeProcessDelay.Milliseconds() > (time.Now().UnixMilli() - int64(createTime))` check simpler and makes mocking easier.

We need this wrapper for the `TestScrape` tests (these used to mock bootTime, but now they mock createTime) since these don't have easy access to the `p.CreateTime` method

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go (hunk 5 above)
Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go]: It was a mistake during rebasing. Fixed

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go (hunk 6 above)
Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go]: It can work as it was previously. Undid the change

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go (hunk 8 above)
Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: Rebase issue. Fixed

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go [hunk 9]
```diff
@@ -306,6 +306,7 @@ func (p *processHandleMock) IOCounters() (*process.IOCountersStat, error) {
 	return args.Get(0).(*process.IOCountersStat), args.Error(1)
 }
 
+
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go:309]: Please make sure no unrelated/accidental changes are done in the PR

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go (hunk 7 above)
Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:145]: How does it make it easier? Doesn't `p.CreateTime` already return `int64` that you want in that line?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go (hunk 9 above)
Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go]: done

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go (hunk 7 above)
Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: Switched to using pcommon.Timestamp

Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: I was actually asking to remove the wrapper and use just `p.CreateTime` for simplicity and consistency with other stats getters that we have in the scraper

Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: We need this wrapper for the `TestScrape` tests (these used to mock bootTime, but now they mock createTime) since these don't have easy access to the `p.CreateTime` method. Here is the associated snippet

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/4aaa26dea8d9eb639752e8904ec96d38d5898d31/receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_test.go#L81-L124

Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: I understand, but why do you need the wrapper? You can mock `p.CreateTime` which has `func(p processHandle) (uint64, error)` signature

Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: How would you mock `p.CreateTime` in the linked test case? I think it is not easily doable so perhaps I am missing something

Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: The same way as you mocking it currently, just change `pcommon.Timestamp` to `uint64`

```go
scraper.getProcessCreateTime = func(p processHandle) (uint64, error) { return createTime, nil } 
```

Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: It was originally set to `uint64` but was changed to pcommon.Timestamp due to the first comment in the chain. Do you want me to change it to how it was then? That is

```
func getProcessCreateTimeInternal(p processHandle) (uint64, error) {
	v, e := p.CreateTime()
	return uint64(v), e
}
```

Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: Or do you mean that it should match the createtime type exactly? That is 

```
func getProcessCreateTimeInternal(p processHandle) (int64, error) {
	v, e := p.CreateTime()
	return v, e
}
```

Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: You don't need any wrappers. I was assuming that p.CreateTime returns `(uint64, error)`. If it's `(int64, error)`, then the signature you need to mock is:

```go
func(p processHandle) (int64, error)
```

Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]:
```
func getProcessCreateTimeInternal(p processHandle) (int64, error) {
	return p.CreateTime()
}
```

is the same as

```go
var getProcessCreateTimeInternal = p.CreateTime
```

Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: > It was originally set to uint64 but was changed to pcommon.Timestamp due to the first comment in the chain.

My original comment was to remove the redundant wrapper

Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: Where would you insert the `var getProcessCreateTimeInternal = p.CreateTime` though? The `newProcessScraper` function doesn't have access to a `processHandle` `p`

Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: right, just use `processHandle.CreateTime`

Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: done

Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: Can you remove the wrapper now?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go [hunk 10]
```diff
@@ -137,3 +138,4 @@ func parentPid(handle processHandle, pid int32) (int32, error) {
 
 	return parent.Pid, nil
 }
+
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:141]: Please avoid accidental changes

**Review by dmitryax — COMMENTED** [review]
Please rebase and make sure CI checks pass

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go (hunk 10 above)
Comment (hsmatulisgoogle) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: sorry, fixed

**Comment by hsmatulisgoogle** [comment]
Rebased, but I need approval to run CI

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
@hsmatulisgoogle please fix linter issues

**Comment by hsmatulisgoogle** [comment]
Fixed the lint errors, but I need approval to run CI
